src/printer-markdown.js

```diff
@@ -625,10 +625,14 @@ function printUrl(url, dangerousCharOrChars) {
     : url;
 }
 
-function printTitle(title) {
-  return title.includes('"') && !title.includes("'")
-    ? `'${title}'`
-    : `"${title.replace(/"/g, '\\"')}"`;
+function printTitle(title, options) {
+  if (!title) {
+    return "";
+  }
+  if (title.includes('"') && title.includes("'") && !title.includes(")")) {
+    return ` (${title})`; // avoid escaped quotes
+  }
+  return " " + util.printString(title, options, { hasQuotes: false });
```

Comment:
Are the escaping rules the same as for JS so that it is safe to reuse `util.printString`?

Reply:
Since `util.printString` requires a raw content but `remark` returns a parsed title, I think it'd be better to use markdown-specific `printTitle`:

- if there're `"` and `'` but no `)`, use `()`-style
- else if `"` is less than `'`, use `"`-style
- else if `'` is less than `"`, use `'`-style
- else use `options.singleQuote`-style

+ escape the char used by the current style
+ if there's a `\` in the parsed title
  - if it's followed by the current style char or `\`, or it's the last char, print `\\`
  - else print `\`

Reply:
Some test cases:

```md

[a](https://example.com "\"")
[a](https://example.com '\"')
[a](https://example.com (\"))

[a](https://example.com "\'")
[a](https://example.com '\'')
[a](https://example.com (\'))

[a](https://example.com "\'")
[a](https://example.com '\)')
[a](https://example.com (\)))

<!-- ↓ mis-parsing, backslashes are missing -->

[a](https://example.com "\\\"")
[a](https://example.com '\\\'')
[a](https://example.com (\\\)))

[a](https://example.com "\\'")
[a](https://example.com '\\"')
[a](https://example.com (\\"))
```

But it seems [remark mis-parsed some escapes](http://astexplorer.net/#/gist/13d926f387a2c0f8b0de6bb0201229fa/e3009dfb87dc663de45cc3df2abd2e8386376cc0) again..., [see CommonMark playground](<http://spec.commonmark.org/dingus/?text=%5Ba%5D(https%3A%2F%2Fexample.com%20%22%5C%22%22)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%27%5C%22%27)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20(%5C%22))%0A%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%22%5C%27%22)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%27%5C%27%27)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20(%5C%27))%0A%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%22%5C%27%22)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%27%5C)%27)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20(%5C)))%0A%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%22%5C%5C%5C%22%22)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%27%5C%5C%5C%27%27)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20(%5C%5C%5C)))%0A%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%22%5C%5C%27%22)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20%27%5C%5C%22%27)%0A%5Ba%5D(https%3A%2F%2Fexample.com%20(%5C%5C%22))>).